cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java

```diff
@@ -0,0 +1,166 @@
+        .build(new CacheLoader<byte[], Long>() {
+
+          @Override
+          public Long load(byte[] topicBytes) throws Exception {
+            Get get = new Get(topicBytes);
+            Result result = metadataTable.get(get);
+            byte[] properties = result.getValue(COL_FAMILY, COL);
+            Map<String, String> propertyMap = GSON.fromJson(Bytes.toString(properties), MAP_TYPE);
+            return TimeUnit.SECONDS.toMillis(Long.parseLong(propertyMap.get(MessagingUtils.Constants.TTL_KEY)));
+          }
+        });
+    }
+  }
+
+  @Override
+  public InternalScanner preFlushScannerOpen(ObserverContext<RegionCoprocessorEnvironment> c, Store store,
+                                             KeyValueScanner memstoreScanner, InternalScanner s) throws IOException {
+    LOG.info("preFlush, filter using TTLFilter");
+    Scan scan = new Scan();
+    scan.setFilter(new TTLFilter(c.getEnvironment(), System.currentTimeMillis()));
+    return new StoreScanner(store, store.getScanInfo(), scan, Collections.singletonList(memstoreScanner),
+                            ScanType.COMPACT_DROP_DELETES, store.getSmallestReadPoint(), HConstants.OLDEST_TIMESTAMP);
+  }
+
+  @Override
+  public InternalScanner preCompactScannerOpen(ObserverContext<RegionCoprocessorEnvironment> c, Store store,
+                                               List<? extends KeyValueScanner> scanners, ScanType scanType,
+                                               long earliestPutTs, InternalScanner s,
+                                               CompactionRequest request) throws IOException {
+    LOG.info("preCompact, filter using TTLFilter");
+    Scan scan = new Scan();
+    scan.setFilter(new TTLFilter(c.getEnvironment(), System.currentTimeMillis()));
+    return new StoreScanner(store, store.getScanInfo(), scan, scanners, scanType, store.getSmallestReadPoint(),
+                            earliestPutTs);
+  }
+
+  private final class TTLFilter extends FilterBase {
+    private final RegionCoprocessorEnvironment env;
+    private final long timestamp;
+
+    private byte[] prevTopicIdBytes;
+    private Long currentTTL;
+
+    TTLFilter(RegionCoprocessorEnvironment env, long timestamp) {
+      this.env = env;
+      this.timestamp = timestamp;
+    }
+
+    @Override
+    public ReturnCode filterKeyValue(Cell cell) throws IOException {
+      int rowKeyOffset = cell.getRowOffset() + PREFIX_LENGTH;
+      int sizeOfRowKey = cell.getRowLength() - PREFIX_LENGTH;
+      long publishTimestamp = MessagingUtils.getPublishTimestamp(cell.getRowArray(), rowKeyOffset, sizeOfRowKey);
+      byte[] topicIdBytes = Arrays.copyOfRange(cell.getRowArray(), rowKeyOffset,
+                                               rowKeyOffset + MessagingUtils.getTopicLengthMessageEntry(sizeOfRowKey));
+      try {
+        if (prevTopicIdBytes == null || (!Arrays.equals(prevTopicIdBytes, topicIdBytes)) || currentTTL == null) {
+          currentTTL = topicCache.get(topicIdBytes);
```

Comment:
`byte[]` doesn't support `equals` and `hashCode` as required by `LoadingCache` (which is a `map`), meaning this call will always try to load the topic info. Better use a `ByteBuffer` to wrap row array with offset and len as `ByteBuffer` supports `equals` and `hashCode` correctly.